mitodl/mitxonline
# Adds bulk code generation to staff-dashboard

#### Pre-Flight checklist


#### What are the relevant tickets?

Closes #1623

#### What's this PR do?

* Ups the maximum discount code length to 100 characters (since bulk codes use UUIDs, we were left with 13 characters for the bulk code prefix, which has been sort of a problem)
* Refactors bulk code generation into `api.py`
* Adds a Create Bulk Discount form to the staff dashboard to allow staff users to create codes on their own
   * The form works like the existing Create Discount form, with a couple of small modifications to support bulk generation
   * Generated codes are displayed after creation and the option is given to download the codes as a CSV file

#### How should this be manually tested?

Note that both your watch and refine containers should probably be restarted for testing this as there are new JavaScript dependencies. 

1. Load the staff dashboard and go to Discounts.
2. Click the Bulk Create button. 
3. Fill out the form appropriately. 
   4. Note that the prefix does not automatically append a separator between it and the code itself, so if you want one (our convention is to use a `-`) you need to specify it explicitly.
   5. If you fill out the date fields, make sure to also specify a time. Times are in UTC.
6. Click Save to create.
7. You should see the list of discount codes. 
8. Click Download CSV to download the generated codes in a CSV file.
9. The CSV file should be correct.

#### Screenshots (if appropriate)

![Screenshot 2023-05-24 at 12 36 38 PM](https://github.com/mitodl/mitxonline/assets/945611/ab1c8be0-2ee3-466f-b8e5-e2402d4751c1)
![Screenshot 2023-05-24 at 12 36 04 PM](https://github.com/mitodl/mitxonline/assets/945611/4ebca8d5-05be-4b9e-b2b9-d50967f85371)
![Screenshot 2023-05-24 at 12 35 27 PM](https://github.com/mitodl/mitxonline/assets/945611/d11049bc-4858-4239-8e50-e36b6928cd48)

(I changed the button title from Create Batch to Bulk Create for consistency but did not redo the screenshot.)
